File: nodejs/04_vision.js
/**
 * ============================================================================
 * Example 3: Vision - Teaching AI to "See" Images
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 * - How to send images to AI models for analysis
 * - Converting images to base64 format for API transmission
 * - Working with file system operations in Node.js
 * - Multi-modal AI (text + images together)
 *
 * WHAT YOU'LL LEARN:
 * - Reading image files from disk
 * - Base64 encoding explained and why it's needed
 * - Creating data URLs for image transmission
 * - Detecting file types and MIME types
 * - Handling file system errors gracefully
 *
 * PREREQUISITES:
 * - Completed 01_basic_chat.js and 02_system_prompt.js
 * - DEMETERICS_API_KEY environment variable set
 * - A test image file (or use the curl command below to download one)
 *
 * EXPECTED OUTPUT:
 * - Image file information (size, base64 size)
 * - AI's detailed description of what's in the image
 * - Token usage statistics
 *
 * RUN THIS:
 * # First, create a test image:
 * curl -o test_image.jpg https://picsum.photos/400/300
 * # Then run:
 * node 03_vision.js
 * ============================================================================
 */

// Step 1: Import required modules
const https = require('https');  // For making API requests
const fs = require('fs');        // File System - for reading files
const path = require('path');    // For working with file paths

// Step 2: Define the path to the image you want to analyze
// Using test_image.jpg from the root directory
const imagePath = '../test_image.jpg';

// Step 3: Read the image file and convert it to base64 format
// Base64 is a way to represent binary data (images) as text
let imageBase64;
try {
  // Step 3a: Read the entire image file into memory as a Buffer (binary data)
  const imageBuffer = fs.readFileSync(imagePath);

  // Step 3b: Convert the binary Buffer to a base64 string
  // Base64 encoding allows binary data to be transmitted in JSON
  imageBase64 = imageBuffer.toString('base64');

  // Step 3c: Display image information for educational purposes
  console.log(`Image loaded: ${imagePath}`);
  console.log(`Image size: ${(imageBuffer.length / 1024).toFixed(2)} KB`);
  console.log(`Base64 size: ${(imageBase64.length / 1024).toFixed(2)} KB`);
  console.log('(Base64 is ~33% larger than original)\n');
} catch (error) {
  // Step 3d: Handle file not found or other errors
  console.error(`Error reading image: ${error.message}`);
  console.log('\nTo create a test image, run:');
  console.log('  curl -o test_image.jpg https://picsum.photos/400/300');
  process.exit(1);  // Exit with error code
}

// Step 4: Detect the image type from file extension
// This determines the correct MIME type for the data URL
const ext = path.extname(imagePath).toLowerCase();  // e.g., ".jpg"
const mimeTypes = {
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.png': 'image/png',
  '.gif': 'image/gif',
  '.webp': 'image/webp'
};
const mimeType = mimeTypes[ext] || 'image/jpeg';  // Default to jpeg if unknown

// Step 5: Build the data URL
// Format: data:<mime-type>;base64,<base64-encoded-data>
const imageUrl = `data:${mimeType};base64,${imageBase64}`;

// Step 6: Create the request payload with both text and image
// Notice the content is now an ARRAY of objects (not a string)
const data = JSON.stringify({
  model: "meta-llama/llama-4-scout-17b-16e-instruct",
  messages: [
    {
      role: "user",
      content: [  // Content is an array for multi-modal input
        {
          type: "text",  // First element: the text prompt
          text: "What is in this image? Describe it in detail."
        },
        {
          type: "image_url",  // Second element: the image
          image_url: {
            url: imageUrl  // The data URL we created above
          }
        }
      ]
    }
  ],
  temperature: 0.3,  // Low temperature for consistent descriptions
  max_tokens: 500    // More tokens needed for detailed image descriptions
});

// Step 7: Configure the HTTPS request options
const options = {
  hostname: 'api.demeterics.com',
  port: 443,
  path: '/groq/v1/chat/completions',
  method: 'POST',
  headers: {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${process.env.DEMETERICS_API_KEY}`,
    'Content-Length': Buffer.byteLength(data)
  }
};

// Step 8: Create and send the HTTPS request
const req = https.request(options, (res) => {
  // Step 9: Accumulate response data
  let responseData = '';

  // Step 10: Handle incoming data chunks
  res.on('data', (chunk) => {
    responseData += chunk;
  });

  // Step 11: Process the complete response
  res.on('end', () => {
    try {
      // Step 12: Parse the JSON response
      const parsed = JSON.parse(responseData);

      // Step 13: Display the AI's vision analysis
      console.log('AI Vision Analysis:');
      console.log('='.repeat(50));
      console.log(parsed.choices[0].message.content);
      console.log('='.repeat(50));

      // Step 14: Show token usage
      console.log('\nToken Usage:');
      console.log(`Total: ${parsed.usage.total_tokens} tokens`);

    } catch (error) {
      // Step 15: Handle parsing errors
      console.error('Error:', error.message);

      // Check for common image-related errors
      if (responseData.includes('image')) {
        console.log('\nImage might be too large. Try:');
        console.log('- Resize to under 4MB (for base64)');
        console.log('- Use JPEG instead of PNG');
        console.log('- Reduce image quality/resolution');
      }
      console.log('\nRaw response:', responseData);
    }
  });
});

// Step 16: Handle request errors
req.on('error', (error) => {
  console.error('Request error:', error.message);
});

// Step 17: Send the request data (includes the base64 image)
req.write(data);

// Step 18: Complete and send the request
req.end();

// Node.js file system concepts:
//
// require('fs')
//   File System module - for reading/writing files
//
// require('path')
//   Path module - for working with file paths
//
// fs.readFileSync(path)
//   Reads entire file into memory (synchronous = waits)
//   Returns a Buffer (binary data)
//   Use fs.readFile() for async (non-blocking)
//
// buffer.toString('base64')
//   Converts binary data to base64 text
//   Other options: 'utf8', 'hex', 'ascii'
//
// path.extname(filepath)
//   Gets file extension: '/path/to/photo.jpg' → '.jpg'
//
// process.exit(1)
//   Exits program with error code 1
//   0 = success, non-zero = error
//
// Base64 encoding:
//   Binary data → Text representation
//   Required because JSON can't contain raw binary
//   Size increase: ~33% larger than original
//   Formula: base64_size = Math.ceil(file_size * 4/3)
//
// Image size limits:
//   Base64 (embedded): 4 MB max
//   URL (remote): 20 MB max
//   Resolution: 33 megapixels max
//
// Content array structure:
//   [
//     { type: "text", text: "..." },        // Your question
//     { type: "image_url", image_url: {...} }  // The image
//   ]
//   Order doesn't matter - AI sees both
//
// Try different questions:
//   "What colors are in this image?"
//   "Is there any text? If so, what does it say?" (OCR)
//   "Describe this image for a blind person"
//   "What's the mood or emotion of this image?"

File: cpp/04_vision.cpp
/**
 * ============================================================================
 * Example 4: Vision - Image Analysis with AI
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 *   - Sending images to AI vision models
 *   - Base64 encoding images in C++
 *   - Multimodal (text + image) API requests
 *
 * WHAT YOU'LL LEARN:
 *   - How vision models process images
 *   - Base64 encoding with modern C++
 *   - File I/O with std::ifstream
 *   - Building complex nested JSON
 *
 * COMPILE:
 *   g++ -std=c++17 -o 04_vision 04_vision.cpp -lcurl -I lib/
 *
 * RUN:
 *   ./04_vision
 *
 * NOTE: Requires test_image.jpg in the parent directory
 *
 * ============================================================================
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <cstdlib>
#include <curl/curl.h>
#include "json.hpp"

using json = nlohmann::json;

/* Callback for curl to write response data */
static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    std::string *response = static_cast<std::string*>(userp);
    response->append(static_cast<char*>(contents), size * nmemb);
    return size * nmemb;
}

/* ============================================================================
 * Base64 encoding function
 * ============================================================================
 * Base64 converts binary data to ASCII text using 64 safe characters.
 * This is necessary because JSON can't contain raw binary data.
 */
std::string base64_encode(const std::vector<unsigned char> &data) {
    static const char table[] =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    std::string result;
    result.reserve(4 * ((data.size() + 2) / 3));

    size_t i = 0;
    while (i < data.size()) {
        uint32_t octet_a = i < data.size() ? data[i++] : 0;
        uint32_t octet_b = i < data.size() ? data[i++] : 0;
        uint32_t octet_c = i < data.size() ? data[i++] : 0;

        uint32_t triple = (octet_a << 16) + (octet_b << 8) + octet_c;

        result += table[(triple >> 18) & 0x3F];
        result += table[(triple >> 12) & 0x3F];
        result += table[(triple >> 6) & 0x3F];
        result += table[triple & 0x3F];
    }

    /* Add padding */
    size_t mod = data.size() % 3;
    if (mod > 0) {
        result[result.size() - 1] = '=';
        if (mod == 1) {
            result[result.size() - 2] = '=';
        }
    }

    return result;
}

/* Read entire file into a vector */
std::vector<unsigned char> read_file(const std::string &filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + filename);
    }

    /* Read file into vector */
    return std::vector<unsigned char>(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    );
}

int main() {
    /* Step 1: Get API key */
    const char *api_key_cstr = std::getenv("DEMETERICS_API_KEY");
    if (api_key_cstr == nullptr) {
        std::cerr << "Error: DEMETERICS_API_KEY not set\n";
        return 1;
    }
    std::string api_key(api_key_cstr);

    /* Step 2: Read and encode image file */
    std::string image_path = "../test_image.jpg";
    std::cout << "Reading image: " << image_path << "\n";

    std::vector<unsigned char> image_data;
    try {
        image_data = read_file(image_path);
    } catch (const std::exception &e) {
        std::cerr << "Error: " << e.what() << "\n";
        std::cerr << "Make sure test_image.jpg exists in the parent directory\n";
        return 1;
    }

    std::cout << "Image size: " << image_data.size() << " bytes\n";

    /* Step 3: Base64 encode the image */
    std::string image_base64 = base64_encode(image_data);
    std::cout << "Base64 length: " << image_base64.size() << " characters\n\n";

    /* Step 4: Initialize curl */
    curl_global_init(CURL_GLOBAL_ALL);
    CURL *curl = curl_easy_init();
    if (curl == nullptr) {
        curl_global_cleanup();
        return 1;
    }

    /* ========================================================================
     * Step 5: Build JSON request with image
     * ========================================================================
     * Vision models accept images as base64-encoded data URLs.
     * Format: "data:image/jpeg;base64,<base64_data>"
     *
     * nlohmann/json makes this nested structure easy!
     */

    std::string data_url = "data:image/jpeg;base64," + image_base64;

    json request_body = {
        {"model", "meta-llama/llama-4-scout-17b-16e-instruct"},
        {"messages", {
            {
                {"role", "user"},
                {"content", {
                    /* Text part: the question about the image */
                    {
                        {"type", "text"},
                        {"text", "What do you see in this image? Describe it in detail."}
                    },
                    /* Image part: the base64-encoded image */
                    {
                        {"type", "image_url"},
                        {"image_url", {{"url", data_url}}}
                    }
                }}
            }
        }},
        {"max_tokens", 500}
    };

    std::string json_payload = request_body.dump();

    /* Step 6: Set up headers */
    struct curl_slist *headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    std::string auth_header = "Authorization: Bearer " + api_key;
    headers = curl_slist_append(headers, auth_header.c_str());

    /* Step 7: Configure and perform request */
    std::string response_data;
    curl_easy_setopt(curl, CURLOPT_URL, "https://api.demeterics.com/groq/v1/chat/completions");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_payload.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);

    std::cout << "Analyzing image with AI vision...\n\n";
    CURLcode res = curl_easy_perform(curl);

    if (res != CURLE_OK) {
        std::cerr << "Error: " << curl_easy_strerror(res) << "\n";
    } else {
        /* Step 8: Parse and display response */
        try {
            json response = json::parse(response_data);

            if (response.contains("choices") && !response["choices"].empty()) {
                std::string answer = response["choices"][0]["message"]["content"];
                std::cout << "AI Vision Analysis:\n";
                std::cout << "==================\n";
                std::cout << answer << "\n";
            }

            /* Display token usage */
            if (response.contains("usage")) {
                std::cout << "\nToken Usage:\n";
                std::cout << "  Prompt: " << response["usage"]["prompt_tokens"] << "\n";
                std::cout << "  Response: " << response["usage"]["completion_tokens"] << "\n";
            }
        } catch (const json::exception &e) {
            std::cerr << "Error parsing response: " << e.what() << "\n";
            std::cerr << "Raw response: " << response_data << "\n";
        }
    }

    /* Step 9: Clean up */
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    curl_global_cleanup();

    return 0;
}

/*
 * VISION API NOTES:
 *
 * 1. IMAGE FORMATS:
 *    - JPEG, PNG, GIF, WebP are supported
 *    - Use appropriate MIME type in data URL
 *
 * 2. IMAGE SIZE:
 *    - Large images are automatically resized
 *    - More pixels = more tokens = higher cost
 *    - Consider resizing before sending
 *
 * 3. DATA URL FORMAT:
 *    data:<mime_type>;base64,<base64_data>
 *    Example: data:image/jpeg;base64,/9j/4AAQ...
 *
 * 4. ALTERNATIVE: URL
 *    Instead of base64, you can provide a URL:
 *    {"type": "image_url", "image_url": {"url": "https://..."}}
 *
 * C++ CONCEPTS USED:
 *
 * std::vector<unsigned char>
 *   Dynamic array for binary data. Manages its own memory.
 *
 * std::ifstream with std::ios::binary
 *   Binary file input. Essential for images!
 *
 * std::istreambuf_iterator
 *   Iterator to read file contents into a container.
 *   Elegant one-liner to read entire file!
 *
 * throw / try / catch
 *   Exception handling. Cleaner than error codes.
 *
 * string::reserve()
 *   Pre-allocate memory for better performance.
 */
